# Phase 2 — Data Understanding

## 2.2 Workbook Structure Inspection

This notebook documents the initial investigation of the raw
UCI Online Retail II dataset before any data cleaning or transformation.

### Objective

Understanding the physical structure of the source workbook, including:

- worksheet names
- worksheet dimensions
- column structure
- sample records
- initial structural observations

The raw source file is kept unchanged in `data/raw/`.

In [1]:
from pathlib import Path
import pandas as pd

In [2]:
#Data Checking

raw_file = Path("../data/raw/online_retail_II.xlsx")

raw_file

raw_file.exists()

True

In [3]:
#Data Inspection

excel_file = pd.ExcelFile(raw_file)

print(f"Sheet Names: {excel_file.sheet_names}")
print(f"Number of Sheets: {len(excel_file.sheet_names)}")

Sheet Names: ['Year 2009-2010', 'Year 2010-2011']
Number of Sheets: 2


## 2.2.2 Worksheet Dimensions and Column Structure

The workbook contains two worksheets representing separate reporting
periods. Each worksheet will be inspected independently to determine
its dimensions and column structure before the transaction data is
loaded for detailed analysis.

In [4]:
for sheet in excel_file.sheet_names:
    preview = pd.read_excel(
        raw_file,
        sheet_name=sheet,
        nrows=5
    )
    
    print(f"\n{'=' * 60}")
    print(f"Sheet: {sheet}")
    print(f"{'=' * 60}")
    print(f"Columns: {list(preview.columns)}")
    print(f"Number of columns: {len(preview.columns)}")
    display(preview)


Sheet: Year 2009-2010
Columns: ['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'Price', 'Customer ID', 'Country']
Number of columns: 8


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085,United Kingdom



Sheet: Year 2010-2011
Columns: ['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'Price', 'Customer ID', 'Country']
Number of columns: 8


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom


### Initial Structural Observations

- The workbook contains two worksheets representing separate reporting periods.
- Both worksheets contain the same 8 columns.
- The observed column structure is consistent across the two worksheets.
- The `StockCode` field contains both numeric-looking and alphanumeric product identifiers.
- An invoice can appear across multiple rows, indicating that the data is recorded at transaction-line level rather than one row per invoice.
- No data-cleaning decisions have been made at this stage; these are observations from the raw source.

## 2.2.3 Worksheet Dimensions

Determine the number of rows and columns in each worksheet before
performing detailed data-quality analysis.

In [5]:
for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )
    
    print(f"\n{'=' * 60}")
    print(f"Sheet: {sheet}")
    print(f"{'=' * 60}")
    print(f"Rows: {data.shape[0]:,}")
    print(f"Columns: {data.shape[1]}")


Sheet: Year 2009-2010
Rows: 525,461
Columns: 8

Sheet: Year 2010-2011
Rows: 541,910
Columns: 8


### Dimension Findings

The raw workbook contains:

- **Year 2009-2010:** 525,461 rows and 8 columns.
- **Year 2010-2011:** 541,910 rows and 8 columns.
- **Combined:** 1,067,371 transaction-line records across 8 columns.

The rows represent transaction-line records rather than confirmed unique
orders. An invoice may appear across multiple rows because an invoice can
contain multiple products.

No cleaning or exclusion decisions have been made at this stage.

## 2.2.4 Initial Data Types

The next step is to inspect the data types assigned to each field when
the raw worksheets are loaded. This is an observation step only; no
type conversions or cleaning will be performed yet.

In [6]:
for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    print(f"\n{'=' * 60}")
    print(f"Sheet: {sheet}")
    print(f"{'=' * 60}")

    print(data.dtypes)


Sheet: Year 2009-2010
Invoice                object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
Price                 float64
Customer ID           float64
Country                object
dtype: object

Sheet: Year 2010-2011
Invoice                object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
Price                 float64
Customer ID           float64
Country                object
dtype: object


In [7]:
for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    print(f"\n{'=' * 60}")
    print(f"Sheet: {sheet}")
    print(f"{'=' * 60}")

    data.info()


Sheet: Year 2009-2010
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[ns]
 5   Price        525461 non-null  float64       
 6   Customer ID  417534 non-null  float64       
 7   Country      525461 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 32.1+ MB

Sheet: Year 2010-2011
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541910 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      541910 non-null  object        
 1

## 2.2.5 Initial Data Dictionary

The following data dictionary describes the fields observed in the raw
workbook. Definitions at this stage are based on the field names,
observed values and transaction structure. Detailed business rules will
be established during later phases.

| Field | Observed Type | Business Role | Initial Interpretation |
|---|---|---|---|
| `Invoice` | object | Transaction identifier | Identifies an invoice/transaction |
| `StockCode` | object | Product identifier | Identifies the product associated with the transaction line |
| `Description` | object | Product attribute | Text description of the product |
| `Quantity` | int64 | Transaction measure | Number of units recorded on the transaction line |
| `InvoiceDate` | datetime64[ns] | Time dimension | Date and time associated with the invoice |
| `Price` | float64 | Financial measure | Recorded price associated with the transaction line |
| `Customer ID` | float64 | Customer identifier | Identifies the customer associated with the transaction |
| `Country` | object | Geographic dimension | Country associated with the transaction |

### Initial Schema Observations

- Both reporting-period worksheets have identical column names and observed data types.
- `Quantity` is represented as an integer.
- `Price` is represented as a floating-point value.
- `InvoiceDate` is already recognized as a datetime field.
- `StockCode` is stored as an object field and includes alphanumeric identifiers.
- `Description` contains missing values in both reporting periods.
- `Customer ID` contains substantial missingness in both reporting periods.
- No data transformations or cleaning decisions have been applied.

## 2.2.6 Missing-Value Profile

A missing-value profile is used to quantify the completeness of each
field in the raw transaction data.

This analysis is descriptive only. No records will be removed or
imputed at this stage.

### Missing-Value Findings

The raw data contains missing values in [fields].

The most significant observed missingness occurs in [field], with
approximately [X]% of records missing in [period] and [Y]% in
[period].

These findings indicate that customer-level and/or product-level
analysis may require explicit handling rules during Data Preparation.

No records have been removed or imputed during this stage.

In [ ]:
missing_profiles = []

for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    profile = pd.DataFrame({
        "Column": data.columns,
        "Total Records": len(data),
        "Missing Records": data.isna().sum().values
    })

    profile["Missing %"] = (
        profile["Missing Records"] / profile["Total Records"] * 100
    ).round(2)

    profile["Present Records"] = (
        profile["Total Records"] - profile["Missing Records"]
    )

    profile["Present %"] = (
        profile["Present Records"] / profile["Total Records"] * 100
    ).round(2)

    profile.insert(0, "Reporting Period", sheet)

    missing_profiles.append(profile)

missing_profile = pd.concat(
    missing_profiles,
    ignore_index=True
)

missing_profile

missing_profile.sort_values(
    ["Reporting Period", "Missing %"],
    ascending=[True, False]
)


,Reporting Period,Column,Total Records,Missing Records,Missing %,Present Records,Present %
6,Year 2009-2010,Customer ID,525461,107927,20.54,417534,79.46
2,Year 2009-2010,Description,525461,2928,0.56,522533,99.44
0,Year 2009-2010,Invoice,525461,0,0.00,525461,100.00
1,Year 2009-2010,StockCode,525461,0,0.00,525461,100.00
3,Year 2009-2010,Quantity,525461,0,0.00,525461,100.00
4,Year 2009-2010,InvoiceDate,525461,0,0.00,525461,100.00
5,Year 2009-2010,Price,525461,0,0.00,525461,100.00
7,Year 2009-2010,Country,525461,0,0.00,525461,100.00
14,Year 2010-2011,Customer ID,541910,135080,24.93,406830,75.07
10,Year 2010-2011,Description,541910,1454,0.27,540456,99.73


### Missing-Value Findings

The initial missing-value profile identified two fields with missing
records across both reporting periods:

- `Customer ID` has the highest level of missingness, with 20.54% of
  records missing in 2009-2010 and 24.93% missing in 2010-2011.
- `Description` has comparatively low missingness, with 0.56% missing
  in 2009-2010 and 0.27% missing in 2010-2011.

The fields `Invoice`, `StockCode`, `Quantity`, `InvoiceDate`, `Price`,
and `Country` contain no missing values in either reporting period.

At this stage, no records are removed and no missing values are
imputed. The missing fields will be investigated further to determine
their business impact and appropriate treatment during the Data
Preparation phase.

### Initial Data-Quality Implications

The missing `Customer ID` values may limit customer-level analysis,
but the associated transaction records may still contain useful
information for product, geographic, transaction, and revenue analysis.

Missing `Description` values require further investigation before a
cleaning rule is established.

Missing-value analysis alone does not establish overall data quality.
Additional profiling is required for duplicate records, cancellations
and returns, negative or zero quantities, unusual prices, transaction
identifiers, and categorical inconsistencies.

## 2.2.7 Duplicate & Transaction-Level Investigation

Repeated invoice numbers are not automatically treated as duplicates
because a single invoice may contain multiple product lines.

The objective of this investigation is to distinguish normal
transaction structure from potentially duplicated records.

In [ ]:
duplicate_profiles = []

for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    exact_duplicates = data.duplicated().sum()
    unique_invoices = data["Invoice"].nunique()
    total_records = len(data)

    duplicate_profiles.append({
        "Reporting Period": sheet,
        "Total Records": total_records,
        "Unique Invoices": unique_invoices,
        "Repeated Invoice Records": total_records - unique_invoices,
        "Exact Duplicate Rows": exact_duplicates
    })

duplicate_profile = pd.DataFrame(duplicate_profiles)

duplicate_profile

,Reporting Period,Total Records,Unique Invoices,Repeated Invoice Records,Exact Duplicate Rows
0,Year 2009-2010,525461,28816,496645,6865
1,Year 2010-2011,541910,25900,516010,5268


In [ ]:
invoice_line_profile = []

for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    invoice_counts = data.groupby("Invoice").size()

    invoice_line_profile.append({
        "Reporting Period": sheet,
        "Invoices": len(invoice_counts),
        "Invoices With Multiple Lines": (invoice_counts > 1).sum(),
        "Maximum Lines Per Invoice": invoice_counts.max(),
        "Average Lines Per Invoice": round(invoice_counts.mean(), 2)
    })

invoice_line_profile = pd.DataFrame(invoice_line_profile)

invoice_line_profile

,Reporting Period,Invoices,Invoices With Multiple Lines,Maximum Lines Per Invoice,Average Lines Per Invoice
0,Year 2009-2010,28816,20811,675,18.24
1,Year 2010-2011,25900,20059,1114,20.92


### Duplicate and Transaction-Level Findings

The raw workbook contains repeated invoice numbers across multiple
transaction lines. This is expected because the dataset is structured
at the transaction-line level rather than one row per invoice.

In the 2009-2010 reporting period, 28,816 unique invoices are
represented across 525,461 transaction-line records. In the 2010-2011
reporting period, 25,900 unique invoices are represented across
541,910 transaction-line records.

Most invoices contain multiple transaction lines. The average number of
lines per invoice is 18.24 in 2009-2010 and 20.92 in 2010-2011.

The profile also identified 6,865 exact duplicate rows in 2009-2010 and
5,268 exact duplicate rows in 2010-2011.

Repeated invoice numbers are therefore not treated as duplicates.
Exact duplicate rows require further investigation before a cleaning
rule is established.

### Data-Quality Implications

The transaction grain must be preserved when designing the cleaning
pipeline. Invoice-level aggregation should be performed explicitly
rather than assuming one row represents one order.

Exact duplicate rows are a potential source of double-counting and may
affect revenue, quantity, and other transaction-level metrics. Their
treatment will be determined after further investigation.

## 2.2.8 Returns, Cancellations and Negative Values

E-commerce transaction exports may contain cancelled orders, returns,
adjustments or other non-standard transactions.

This investigation identifies negative and zero quantities and examines
their relationship with invoice identifiers before any treatment rule is
defined.

In [ ]:
transaction_quality_profiles = []

for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    negative_quantity = (data["Quantity"] < 0).sum()
    zero_quantity = (data["Quantity"] == 0).sum()
    negative_price = (data["Price"] < 0).sum()
    zero_price = (data["Price"] == 0).sum()

    cancellation_invoices = (
        data["Invoice"]
        .astype(str)
        .str.upper()
        .str.startswith("C")
        .sum()
    )

    transaction_quality_profiles.append({
        "Reporting Period": sheet,
        "Negative Quantity Rows": negative_quantity,
        "Zero Quantity Rows": zero_quantity,
        "Negative Price Rows": negative_price,
        "Zero Price Rows": zero_price,
        "Cancellation Invoice Rows": cancellation_invoices
    })

transaction_quality_profile = pd.DataFrame(
    transaction_quality_profiles
)

transaction_quality_profile

,Reporting Period,Negative Quantity Rows,Zero Quantity Rows,Negative Price Rows,Zero Price Rows,Cancellation Invoice Rows
0,Year 2009-2010,12326,0,3,3687,10206
1,Year 2010-2011,10624,0,2,2515,9288


In [ ]:
for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    cancellation_mask = (
        data["Invoice"]
        .astype(str)
        .str.upper()
        .str.startswith("C")
    )

    negative_quantity_mask = data["Quantity"] < 0

    print(f"\n{sheet}")
    print("-" * len(sheet))

    print(
        "Negative-quantity rows:",
        negative_quantity_mask.sum()
    )

    print(
        "Negative-quantity rows with cancellation invoice:",
        (
            negative_quantity_mask &
            cancellation_mask
        ).sum()
    )

    print(
        "Negative-quantity rows without cancellation invoice:",
        (
            negative_quantity_mask &
            ~cancellation_mask
        ).sum()
    )


Year 2009-2010
--------------
Negative-quantity rows: 12326
Negative-quantity rows with cancellation invoice: 10205
Negative-quantity rows without cancellation invoice: 2121

Year 2010-2011
--------------
Negative-quantity rows: 10624
Negative-quantity rows with cancellation invoice: 9288
Negative-quantity rows without cancellation invoice: 1336


### Returns, Cancellations and Negative-Value Findings

The transaction profile identified 12,326 negative-quantity rows in
2009-2010 and 10,624 in 2010-2011. No zero-quantity rows were found in
either reporting period.

Most negative-quantity rows are associated with cancellation-style
invoice identifiers. In 2009-2010, 10,205 of 12,326 negative-quantity
rows have cancellation-style invoice identifiers. In 2010-2011,
9,288 of 10,624 negative-quantity rows have cancellation-style invoice
identifiers.

However, 2,121 negative-quantity rows in 2009-2010 and 1,336 in
2010-2011 do not have cancellation-style invoice identifiers. These
records require further investigation before a treatment rule is
defined.

The dataset also contains 3,687 zero-price rows in 2009-2010 and 2,515
in 2010-2011. Five rows contain negative prices across the two periods.

No records are removed at this stage. The findings indicate that the
final reporting model should distinguish positive sales activity from
returns or cancellations rather than treating all negative quantities
as invalid data.

Zero-price and negative-price transactions will be investigated further
before cleaning rules are established.

## 2.2.9 Price Anomaly Investigation

The previous profile identified zero-price and negative-price
transactions.

This investigation examines the affected records and their surrounding
transaction characteristics before any treatment rule is defined.

In [ ]:
for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    negative_price_rows = data[data["Price"] < 0]

    print(f"\n{sheet}")
    print("-" * len(sheet))

    display(negative_price_rows)


Year 2009-2010
--------------


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
179403,A506401,B,Adjust bad debt,1,2010-04-29 13:36:00,-53594.36,NaN,United Kingdom
276274,A516228,B,Adjust bad debt,1,2010-07-19 11:24:00,-44031.79,NaN,United Kingdom
403472,A528059,B,Adjust bad debt,1,2010-10-20 12:04:00,-38925.87,NaN,United Kingdom



Year 2010-2011
--------------


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
299983,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


In [ ]:
zero_price_profiles = []

for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    zero_price = data[data["Price"] == 0]

    zero_price_profiles.append({
        "Reporting Period": sheet,
        "Zero Price Rows": len(zero_price),
        "Unique Invoices": zero_price["Invoice"].nunique(),
        "Unique Products": zero_price["StockCode"].nunique(),
        "Negative Quantity Rows": (zero_price["Quantity"] < 0).sum(),
        "Positive Quantity Rows": (zero_price["Quantity"] > 0).sum(),
        "Missing Customer IDs": zero_price["Customer ID"].isna().sum()
    })

zero_price_profile = pd.DataFrame(zero_price_profiles)

zero_price_profile

,Reporting Period,Zero Price Rows,Unique Invoices,Unique Products,Negative Quantity Rows,Positive Quantity Rows,Missing Customer IDs
0,Year 2009-2010,3687,3346,2195,2121,1566,3656
1,Year 2010-2011,2515,2155,1419,1336,1179,2475


In [ ]:
for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    zero_price = data[data["Price"] == 0]

    print(f"\nTop zero-price products — {sheet}")
    print("-" * 40)

    display(
        zero_price.groupby(
            ["StockCode", "Description"]
        )
        .size()
        .sort_values(ascending=False)
        .head(15)
    )


Top zero-price products — Year 2009-2010
----------------------------------------


StockCode  Description                        
21116      OWL DOORSTOP                           8
22501      PICNIC BASKET WICKER LARGE             7
46000M     POLYESTER FILLER PAD 45x45cm           7
22372      AIRLINE BAG VINTAGE WORLD CHAMPION     6
22469      HEART OF WICKER SMALL                  6
84016      FLAG OF ST GEORGE CAR FLAG             6
22197      SMALL POPCORN HOLDER                   5
22502      PICNIC BASKET WICKER SMALL             5
22624      IVORY KITCHEN SCALES                   5
22428      ENAMEL FIRE BUCKET CREAM               5
22426      ENAMEL WASH BOWL CREAM                 5
22431      WATERING CAN BLUE ELEPHANT             5
46000S     POLYESTER FILLER PAD 40x40cm           5
20852      given away                             4
22465      HANGING METAL STAR LANTERN             4
dtype: int64


Top zero-price products — Year 2010-2011
----------------------------------------


StockCode  Description                    
22676      FRENCH BLUE METAL DOOR SIGN 1      9
22683      FRENCH BLUE METAL DOOR SIGN 8      8
22686      FRENCH BLUE METAL DOOR SIGN No     7
22666      RECIPE BOX PANTRY YELLOW DESIGN    7
22678      FRENCH BLUE METAL DOOR SIGN 3      7
22679      FRENCH BLUE METAL DOOR SIGN 4      7
21116      OWL DOORSTOP                       7
M          Manual                             6
22682      FRENCH BLUE METAL DOOR SIGN 7      6
22625      RED KITCHEN SCALES                 6
22681      FRENCH BLUE METAL DOOR SIGN 6      6
22627      MINT KITCHEN SCALES                6
22680      FRENCH BLUE METAL DOOR SIGN 5      6
22684      FRENCH BLUE METAL DOOR SIGN 9      6
22685      FRENCH BLUE METAL DOOR SIGN 0      5
dtype: int64

### Price Anomaly Findings

The price investigation identified 5 negative-price records across the
two reporting periods. All five records have the description
`Adjust bad debt`, a quantity of 1, no Customer ID, and a United Kingdom
country value. These records appear structurally different from normal
merchandise transactions and should therefore be classified separately
from ordinary product sales rather than automatically treated as
standard sales records.

The profile also identified 6,202 zero-price transaction lines:
3,687 in 2009-2010 and 2,515 in 2010-2011.

Zero-price transactions involve multiple invoices and products and
include both positive and negative quantities. Most also have missing
Customer IDs: 3,656 of 3,687 rows in 2009-2010 and 2,475 of 2,515 rows
in 2010-2011.

The zero-price product descriptions include ordinary merchandise as
well as descriptions such as `given away` and `Manual`. Therefore,
zero-price records cannot be assumed to represent a single business
condition.

### Data-Quality Implications

Negative-price transactions should be investigated as a separate
transaction category because they may represent financial adjustments
rather than merchandise sales.

Zero-price transactions should not be automatically removed or treated
as errors. Their characteristics suggest that the final pipeline should
classify transaction types before calculating sales KPIs.

No cleaning decisions are applied during this investigation phase.

## 2.2.10 Transaction-Type and Product-Code Investigation

The previous investigations identified several records that may
represent non-standard business activity rather than ordinary product
sales.

This investigation examines the structure of StockCode, Invoice and
Description values to identify unusual transaction categories before
cleaning rules are established.

In [ ]:
for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    print(f"\n{sheet}")
    print("=" * 60)

    print("\nInvoice prefix distribution:")
    print(
        data["Invoice"]
        .astype(str)
        .str[0]
        .value_counts()
        .sort_index()
    )

    print("\nStockCode patterns:")
    print(
        data["StockCode"]
        .astype(str)
        .str[0]
        .value_counts()
        .sort_index()
        .head(30)
    )


Year 2009-2010

Invoice prefix distribution:
Invoice
4    108489
5    406763
A         3
C     10206
Name: count, dtype: int64

StockCode patterns:
StockCode
1     10405
2    372529
3     11456
4     16838
5       523
6       351
7     14318
8     89302
9      6642
A        79
B        68
C       139
D       949
G         1
M       850
P       880
S        44
T        17
g        66
m         4
Name: count, dtype: int64

Year 2010-2011

Invoice prefix distribution:
Invoice
5    532619
A         3
C      9288
Name: count, dtype: int64

StockCode patterns:
StockCode
1      7627
2    438444
3      5758
4     11464
5       641
6       113
7      7223
8     62988
9      4656
A        34
B        40
C       160
D       832
M       571
P      1261
S        63
g        34
m         1
Name: count, dtype: int64


In [ ]:
for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    print(f"\n{sheet}")
    print("=" * 60)

    description_counts = (
        data["Description"]
        .fillna("[MISSING]")
        .value_counts()
    )

    print("\nDescriptions appearing most frequently:")
    display(description_counts.head(20))


Year 2009-2010

Descriptions appearing most frequently:


Description
WHITE HANGING HEART T-LIGHT HOLDER    3549
[MISSING]                             2928
REGENCY CAKESTAND 3 TIER              2212
STRAWBERRY CERAMIC TRINKET BOX        1843
PACK OF 72 RETRO SPOT CAKE CASES      1466
ASSORTED COLOUR BIRD ORNAMENT         1457
60 TEATIME FAIRY CAKE CASES           1400
HOME BUILDING BLOCK WORD              1386
JUMBO BAG RED RETROSPOT               1310
LUNCH BAG RED SPOTTY                  1274
REX CASH+CARRY JUMBO SHOPPER          1232
JUMBO STORAGE BAG SUKI                1220
PACK OF 60 PINK PAISLEY CAKE CASES    1196
WOODEN FRAME ANTIQUE WHITE            1190
LUNCH BAG  BLACK SKULL.               1179
BAKING SET 9 PIECE RETROSPOT          1176
LUNCH BAG SUKI  DESIGN                1157
HEART OF WICKER LARGE                 1151
LOVE BUILDING BLOCK WORD              1142
RED HANGING HEART T-LIGHT HOLDER      1129
Name: count, dtype: int64


Year 2010-2011

Descriptions appearing most frequently:


Description
WHITE HANGING HEART T-LIGHT HOLDER    2369
REGENCY CAKESTAND 3 TIER              2200
JUMBO BAG RED RETROSPOT               2159
PARTY BUNTING                         1727
LUNCH BAG RED RETROSPOT               1638
ASSORTED COLOUR BIRD ORNAMENT         1501
SET OF 3 CAKE TINS PANTRY DESIGN      1473
[MISSING]                             1454
PACK OF 72 RETROSPOT CAKE CASES       1385
LUNCH BAG  BLACK SKULL.               1350
NATURAL SLATE HEART CHALKBOARD        1280
POSTAGE                               1253
JUMBO BAG PINK POLKADOT               1251
HEART OF WICKER SMALL                 1237
JAM MAKING SET WITH JARS              1229
JUMBO STORAGE BAG SUKI                1214
PAPER CHAIN KIT 50'S CHRISTMAS        1210
JUMBO SHOPPER VINTAGE RED PAISLEY     1202
LUNCH BAG CARS BLUE                   1197
LUNCH BAG SPACEBOY DESIGN             1192
Name: count, dtype: int64

## 2.2.11 Non-Standard StockCode Investigation

The StockCode field contains both numeric-leading and alphabetic-leading
values.

Because some alphabetic StockCodes may represent operational,
adjustment, or other non-standard transactions, their associated
descriptions, quantities, prices and invoice patterns will be examined
before transaction-classification rules are established.

In [ ]:
for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    data["StockCode_Str"] = data["StockCode"].astype(str).str.strip()

    non_numeric = data[
        ~data["StockCode_Str"].str[0].str.isdigit()
    ].copy()

    profile = (
        non_numeric
        .groupby("StockCode_Str")
        .agg(
            Rows=("StockCode_Str", "size"),
            Descriptions=("Description", "nunique"),
            Negative_Quantity=("Quantity", lambda x: (x < 0).sum()),
            Zero_Price=("Price", lambda x: (x == 0).sum()),
            Positive_Price=("Price", lambda x: (x > 0).sum()),
            Missing_Customer_ID=("Customer ID", lambda x: x.isna().sum())
        )
        .sort_values("Rows", ascending=False)
    )

    print(f"\n{sheet}")
    print("=" * 70)

    display(profile)


Year 2009-2010


,Rows,Descriptions,Negative_Quantity,Zero_Price,Positive_Price,Missing_Customer_ID
StockCode_Str,,,,,,
POST,865,1,103,3,862,43
M,850,1,293,1,849,200
DOT,736,1,2,4,732,736
C2,138,1,5,2,136,13
D,100,1,95,0,100,3
ADJUST,67,3,31,0,67,6
BANK CHARGES,65,2,42,0,65,39
S,41,1,40,0,41,41
DCGS0058,31,1,0,1,30,31



Year 2010-2011


,Rows,Descriptions,Negative_Quantity,Zero_Price,Positive_Price,Missing_Customer_ID
StockCode_Str,,,,,,
POST,1257,1,126,4,1253,60
DOT,710,1,1,3,707,694
M,571,1,244,6,565,106
C2,144,1,2,1,143,10
D,77,1,77,0,77,0
S,63,1,61,0,63,63
BANK CHARGES,37,1,25,0,37,25
AMAZONFEE,34,1,32,0,34,34
CRUK,16,1,16,0,16,0


In [ ]:
for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    data["StockCode_Str"] = data["StockCode"].astype(str).str.strip()

    non_numeric = data[
        ~data["StockCode_Str"].str[0].str.isdigit()
    ].copy()

    display(
        non_numeric[
            [
                "StockCode_Str",
                "Description",
                "Quantity",
                "Price",
                "Invoice",
                "Customer ID",
                "Country"
            ]
        ]
        .drop_duplicates()
        .sort_values("StockCode_Str")
        .head(100)
    )

,StockCode_Str,Description,Quantity,Price,Invoice,Customer ID,Country
71045,ADJUST,Adjustment by john on 26/01/2010 16,-1,13.14,C495761,13664.0,United Kingdom
71053,ADJUST,Adjustment by john on 26/01/2010 16,-1,36.75,C495770,15357.0,Unspecified
71054,ADJUST,Adjustment by john on 26/01/2010 16,-1,13.12,C495771,13776.0,United Kingdom
71055,ADJUST,Adjustment by john on 26/01/2010 16,-1,22.95,C495772,12600.0,Germany
71056,ADJUST,Adjustment by john on 26/01/2010 16,1,4.57,495773,12600.0,Germany
...,...,...,...,...,...,...,...
279647,BANK CHARGES,Bank Charges,-1,165.84,C516612,NaN,United Kingdom
331934,BANK CHARGES,Bank Charges,1,15.00,521809,14911.0,EIRE
143558,BANK CHARGES,Bank Charges,-1,25.00,C503050,NaN,United Kingdom
179427,BANK CHARGES,Bank Charges,-1,25.00,C506405,NaN,United Kingdom


,StockCode_Str,Description,Quantity,Price,Invoice,Customer ID,Country
191386,AMAZONFEE,AMAZON FEE,-1,7006.83,C553355,NaN,United Kingdom
446533,AMAZONFEE,AMAZON FEE,-1,8286.22,C574902,NaN,United Kingdom
312246,AMAZONFEE,AMAZON FEE,-1,6662.51,C564341,NaN,United Kingdom
262413,AMAZONFEE,AMAZON FEE,-1,4383.62,C559915,NaN,United Kingdom
15016,AMAZONFEE,AMAZON FEE,-1,13541.33,C537630,NaN,United Kingdom
...,...,...,...,...,...,...,...
427576,C2,CARRIAGE,1,50.00,573391,14911.0,EIRE
508632,C2,CARRIAGE,1,50.00,579256,13349.0,United Kingdom
50141,C2,CARRIAGE,1,50.00,540547,14911.0,EIRE
360310,C2,CARRIAGE,1,50.00,568221,14911.0,EIRE


## 2.2.11 — Final investigation

We have identified the codes, but we haven't yet established their exact descriptions and business behavior across the whole dataset.

In [ ]:
for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    data["StockCode_Str"] = data["StockCode"].astype(str).str.strip()

    non_numeric = data[
        ~data["StockCode_Str"].str[0].str.isdigit()
    ].copy()

    summary = (
        non_numeric
        .groupby("StockCode_Str")
        .agg(
            Rows=("StockCode_Str", "size"),
            Description=("Description", lambda x: " | ".join(
                x.dropna().astype(str).unique()[:3]
            )),
            Min_Quantity=("Quantity", "min"),
            Max_Quantity=("Quantity", "max"),
            Min_Price=("Price", "min"),
            Max_Price=("Price", "max"),
            Missing_Customer_ID=("Customer ID", lambda x: x.isna().sum()),
            Invoice_Prefixes=("Invoice", lambda x: ", ".join(
                sorted(
                    x.astype(str)
                    .str[0]
                    .dropna()
                    .unique()
                )
            ))
        )
        .sort_values("Rows", ascending=False)
    )

    print(f"\n{sheet}")
    print("=" * 100)

    display(summary)


Year 2009-2010


,Rows,Description,Min_Quantity,Max_Quantity,Min_Price,Max_Price,Missing_Customer_ID,Invoice_Prefixes
StockCode_Str,,,,,,,,
POST,865,POSTAGE,-16,1000,0.000,850.00,43,"4, 5, C"
M,850,Manual,-408,144,0.000,25111.09,200,"4, 5, C"
DOT,736,DOTCOM POSTAGE,-1,500,0.000,1081.70,736,"4, 5, C"
C2,138,CARRIAGE,-1,200,0.000,110.00,13,"4, 5, C"
D,100,Discount,-324,192,0.010,1269.51,3,"5, C"
ADJUST,67,Adjustment by john on 26/01/2010 16 | Adjustme...,-1,1,4.570,5117.03,6,"4, C"
BANK CHARGES,65,Bank Charges | Bank Charges,-11,1,0.960,18910.69,39,"4, 5, C"
S,41,SAMPLES,-1,1,2.880,605.18,41,"5, C"
DCGS0058,31,MISO PRETTY GUM,1,6,0.000,0.95,31,"4, 5"



Year 2010-2011


,Rows,Description,Min_Quantity,Max_Quantity,Min_Price,Max_Price,Missing_Customer_ID,Invoice_Prefixes
StockCode_Str,,,,,,,,
POST,1257,POSTAGE,-4,1000,0.000,8142.750,60,"5, C"
DOT,710,DOTCOM POSTAGE,-1,1000,0.000,4505.170,694,"5, C"
M,571,Manual,-1350,1600,0.000,38970.000,106,"5, C"
C2,144,CARRIAGE,-1,150,0.000,150.000,10,"5, C"
D,77,Discount,-720,-1,0.010,1867.860,0,C
S,63,SAMPLES,-1,1,2.800,570.000,63,"5, C"
BANK CHARGES,37,Bank Charges,-1,1,0.001,1050.150,25,"5, C"
AMAZONFEE,34,AMAZON FEE,-1,1,1.000,17836.460,34,"5, C"
CRUK,16,CRUK Commission,-1,-1,1.600,1100.440,0,C


## 2.2.11 Non-Standard StockCode Investigation

The `StockCode` field contains both numeric-leading and alphabetic-leading
values. Further investigation showed that alphabetic StockCodes do not
represent a single type of data-quality problem.

Several codes are associated with operational or financial transactions.
For example:

- `C2` is associated with `CARRIAGE`.
- `ADJUST` is associated with adjustment records.
- `BANK CHARGES` is associated with bank charge transactions.
- `AMAZONFEE` is associated with Amazon fee transactions.
- `M` represents another non-standard transaction type with a substantial
  number of negative-quantity records.
- `POST` occurs frequently and represents a non-standard transaction
  category requiring separate treatment.
- `DOT`, `DCGS...`, `gift_0001_*`, and other codes also appear in the
  non-standard StockCode population.

Some non-standard codes have substantial missing `Customer ID` values,
while others contain positive or negative quantities and zero or positive
prices.

### Key Finding

Alphabetic or non-numeric `StockCode` values cannot be treated as invalid
product records solely because they are non-numeric.

The investigation indicates that the raw dataset contains multiple
transaction types beyond ordinary merchandise sales, including operational,
financial, shipping, fee, adjustment, gift, and test-like records.

### Data-Quality Implications

The cleaning pipeline should therefore classify transaction records based
on multiple fields rather than removing rows solely according to the
`StockCode` format.

Relevant fields for later transaction classification include:

- `StockCode`
- `Description`
- `Invoice`
- `Quantity`
- `Price`
- `Customer ID`

No records are removed at this stage. Final inclusion/exclusion rules will
be established during the Data Preparation phase after transaction
classification has been completed.

## 2.2.12 — Date & Time Investigation

In [ ]:
for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    print(f"\n{sheet}")
    print("=" * 70)

    print("Minimum InvoiceDate:", data["InvoiceDate"].min())
    print("Maximum InvoiceDate:", data["InvoiceDate"].max())
    print("Missing InvoiceDate:", data["InvoiceDate"].isna().sum())

    monthly = (
        data.assign(
            Month=data["InvoiceDate"].dt.to_period("M")
        )
        .groupby("Month")
        .size()
    )

    print("\nRows by month:")
    display(monthly)


Year 2009-2010
Minimum InvoiceDate: 2009-12-01 07:45:00
Maximum InvoiceDate: 2010-12-09 20:01:00
Missing InvoiceDate: 0

Rows by month:


Month
2009-12    45228
2010-01    31555
2010-02    29388
2010-03    41511
2010-04    34057
2010-05    35323
2010-06    39983
2010-07    33383
2010-08    33306
2010-09    42091
2010-10    59098
2010-11    78015
2010-12    22523
Freq: M, dtype: int64


Year 2010-2011
Minimum InvoiceDate: 2010-12-01 08:26:00
Maximum InvoiceDate: 2011-12-09 12:50:00
Missing InvoiceDate: 0

Rows by month:


Month
2010-12    42481
2011-01    35147
2011-02    27707
2011-03    36748
2011-04    29916
2011-05    37030
2011-06    36874
2011-07    39518
2011-08    35284
2011-09    50226
2011-10    60742
2011-11    84711
2011-12    25526
Freq: M, dtype: int64

## 2.2.12 Date and Time Investigation

The `InvoiceDate` field was investigated to assess temporal coverage,
completeness and monthly transaction distribution.

### Date Coverage

For the `Year 2009-2010` worksheet:

* Earliest `InvoiceDate`: 2009-12-01 07:45:00
* Latest `InvoiceDate`: 2010-12-09 20:01:00
* Missing `InvoiceDate`: 0

For the `Year 2010-2011` worksheet:

* Earliest `InvoiceDate`: 2010-12-01 08:26:00
* Latest `InvoiceDate`: 2011-12-09 12:50:00
* Missing `InvoiceDate`: 0

Both worksheets therefore contain complete `InvoiceDate` values.

### Monthly Transaction Distribution

The transaction-line volume varies considerably across months.

For the `Year 2009-2010` worksheet, monthly records range from
29,388 in February 2010 to 78,015 in November 2010.

For the `Year 2010-2011` worksheet, monthly records range from
25,526 in December 2011 to 84,711 in November 2011.

The higher transaction-line volumes observed around October and November
will be considered during later business analysis and reporting.

### December 2010 Overlap

Both worksheets contain records from December 2010:

* `Year 2009-2010`: 22,523 records
* `Year 2010-2011`: 42,481 records

This overlap should not automatically be treated as duplicated data.
The workbook is divided into reporting-period worksheets, so any potential
transaction overlap must be investigated using transaction-level fields
before a deduplication rule is established.

### Key Finding

`InvoiceDate` is complete across both worksheets and provides sufficient
temporal information for subsequent monthly reporting.

No records are removed and no date values are modified during the Data
Understanding phase.

### Data-Quality Implications

The date field itself does not currently present a missing-value problem.
However, the overlap around December 2010 and the varying monthly
transaction volumes should be considered when the two worksheets are
combined and when monthly reporting logic is developed.


## Next: 2.2.13 — Country & Customer Coverage

In [ ]:
for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    print(f"\n{sheet}")
    print("=" * 70)

    # 1. Country coverage
    print("Unique countries:", data["Country"].nunique())

    print("\nTop countries by transaction-line count:")
    display(
        data["Country"]
        .value_counts()
        .head(15)
    )

    # 2. Customer coverage
    print("\nUnique Customer IDs:", data["Customer ID"].nunique())

    # 3. Missing Customer ID by country
    country_customer = (
        data.groupby("Country")
        .agg(
            Total_Rows=("Customer ID", "size"),
            Missing_Customer_ID=("Customer ID", lambda x: x.isna().sum())
        )
    )

    country_customer["Missing_Rate"] = (
        country_customer["Missing_Customer_ID"]
        / country_customer["Total_Rows"]
        * 100
    )

    print("\nCountries with highest missing Customer ID rates:")
    display(
        country_customer
        .sort_values("Missing_Rate", ascending=False)
        .head(15)
    )

    # 4. Check country formatting
    print("\nMissing Country values:", data["Country"].isna().sum())

    print("\nCountry values with surrounding whitespace:")
    whitespace_country = data[
        data["Country"].astype(str).str.strip() != data["Country"].astype(str)
    ]

    print(len(whitespace_country))


Year 2009-2010
Unique countries: 40

Top countries by transaction-line count:


Country
United Kingdom     485852
EIRE                 9670
Germany              8129
France               5772
Netherlands          2769
Spain                1278
Switzerland          1187
Portugal             1101
Belgium              1054
Channel Islands       906
Sweden                902
Italy                 731
Australia             654
Cyprus                554
Austria               537
Name: count, dtype: int64


Unique Customer IDs: 4383

Countries with highest missing Customer ID rates:


,Total_Rows,Missing_Customer_ID,Missing_Rate
Country,,,
Hong Kong,76,76,100.000000
Bermuda,34,34,100.000000
Lebanon,13,13,100.000000
Bahrain,107,65,60.747664
RSA,111,46,41.441441
United Arab Emirates,432,114,26.388889
United Kingdom,485852,106429,21.905642
EIRE,9670,960,9.927611
Unspecified,310,30,9.677419



Missing Country values: 0

Country values with surrounding whitespace:
0

Year 2010-2011
Unique countries: 38

Top countries by transaction-line count:


Country
United Kingdom     495478
Germany              9495
France               8558
EIRE                 8196
Spain                2533
Netherlands          2371
Belgium              2069
Switzerland          2002
Portugal             1519
Australia            1259
Norway               1086
Italy                 803
Channel Islands       758
Finland               695
Cyprus                622
Name: count, dtype: int64


Unique Customer IDs: 4372

Countries with highest missing Customer ID rates:


,Total_Rows,Missing_Customer_ID,Missing_Rate
Country,,,
Hong Kong,288,288,100.000000
Unspecified,446,202,45.291480
United Kingdom,495478,133600,26.963861
Israel,297,47,15.824916
Bahrain,19,2,10.526316
EIRE,8196,711,8.674963
Switzerland,2002,125,6.243756
Portugal,1519,39,2.567479
France,8558,66,0.771208



Missing Country values: 0

Country values with surrounding whitespace:
0


## 2.2.13 Country and Customer Coverage

The `Country` and `Customer ID` fields were investigated to understand
geographic coverage and customer-data completeness.

### Country Coverage

The `Year 2009-2010` worksheet contains 40 unique country values, while
the `Year 2010-2011` worksheet contains 38.

The United Kingdom represents the largest share of transaction-line
records in both periods:

* 2009-2010: 485,852 transaction lines
* 2010-2011: 495,478 transaction lines

Other frequently represented countries include EIRE, Germany, France,
Spain, the Netherlands, Belgium and Switzerland.

No missing `Country` values were identified in either worksheet.

No country values with leading or trailing whitespace were identified.

### Customer Coverage

The `Year 2009-2010` worksheet contains 4,383 unique non-missing
`Customer ID` values.

The `Year 2010-2011` worksheet contains 4,372 unique non-missing
`Customer ID` values.

However, customer identification is incomplete and varies considerably
by country.

Examples include:

* Hong Kong: 100% of transaction lines have missing `Customer ID` in both
  periods.
* United Kingdom: 21.91% missing in 2009-2010 and 26.96% missing in
  2010-2011.
* EIRE: 9.93% missing in 2009-2010 and 8.67% missing in 2010-2011.
* France: approximately 1% missing or less in both periods.

This demonstrates that missing customer identifiers are not evenly
distributed across the dataset.

### Key Finding

The `Country` field is structurally complete, while the `Customer ID`
field has substantial and uneven missingness.

Customer-level analysis must therefore distinguish between transactions
with an identified customer and transactions where customer attribution
is unavailable.

### Data-Quality Implications

Records with missing `Customer ID` should not be automatically removed.

Removing these records could exclude a substantial portion of the
transaction data and may disproportionately affect certain geographic
segments.

The `Country` field does not currently require missing-value or
whitespace cleaning. However, country-label standardization may be
considered during Data Preparation.

No records are removed or transformed during Data Understanding.


In [ ]:
for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    data["LineValue"] = data["Quantity"] * data["Price"]

    print(f"\n{sheet}")
    print("=" * 70)

    print("Total raw line value:", data["LineValue"].sum())
    print("Minimum line value:", data["LineValue"].min())
    print("Maximum line value:", data["LineValue"].max())

    print("\nLine-value summary:")
    display(
        data["LineValue"].describe()
    )

    print("\nNegative line values:", (data["LineValue"] < 0).sum())
    print("Zero line values:", (data["LineValue"] == 0).sum())
    print("Positive line values:", (data["LineValue"] > 0).sum())

    print("\nTop 15 transactions by absolute line value:")
    display(
        data.loc[
            data["LineValue"].abs().nlargest(15).index,
            [
                "Invoice",
                "StockCode",
                "Description",
                "Quantity",
                "Price",
                "LineValue",
                "Customer ID",
                "Country"
            ]
        ].sort_values(
            "LineValue",
            key=lambda x: x.abs(),
            ascending=False
        )
    )


Year 2009-2010
Total raw line value: 9539484.634
Minimum line value: -53594.36
Maximum line value: 25111.09

Line-value summary:


count    525461.000000
mean         18.154506
std         160.333083
min      -53594.360000
25%           3.750000
50%           9.950000
75%          17.700000
max       25111.090000
Name: LineValue, dtype: float64


Negative line values: 10208
Zero line values: 3687
Positive line values: 511566

Top 15 transactions by absolute line value:


,Invoice,StockCode,Description,Quantity,Price,LineValue,Customer ID,Country
179403,A506401,B,Adjust bad debt,1,-53594.36,-53594.36,NaN,United Kingdom
276274,A516228,B,Adjust bad debt,1,-44031.79,-44031.79,NaN,United Kingdom
403472,A528059,B,Adjust bad debt,1,-38925.87,-38925.87,NaN,United Kingdom
241824,C512770,M,Manual,-1,25111.09,-25111.09,17399.0,United Kingdom
241827,512771,M,Manual,1,25111.09,25111.09,NaN,United Kingdom
320581,C520667,BANK CHARGES,Bank Charges,-1,18910.69,-18910.69,NaN,United Kingdom
432176,530715,84347,ROTATING SILVER ANGELS T-LIGHT HLDR,9360,1.69,15818.40,15838.0,United Kingdom
517953,C537630,AMAZONFEE,AMAZON FEE,-1,13541.33,-13541.33,NaN,United Kingdom
517955,537632,AMAZONFEE,AMAZON FEE,1,13541.33,13541.33,NaN,United Kingdom
519294,C537651,AMAZONFEE,AMAZON FEE,-1,13541.33,-13541.33,NaN,United Kingdom



Year 2010-2011
Total raw line value: 9747765.934
Minimum line value: -168469.6
Maximum line value: 168469.6

Line-value summary:


count    541910.000000
mean         17.987795
std         378.810474
min     -168469.600000
25%           3.400000
50%           9.750000
75%          17.400000
max      168469.600000
Name: LineValue, dtype: float64


Negative line values: 9290
Zero line values: 2515
Positive line values: 530105

Top 15 transactions by absolute line value:


,Invoice,StockCode,Description,Quantity,Price,LineValue,Customer ID,Country
540421,581483,23843,"PAPER CRAFT , LITTLE BIRDIE",80995,2.08,168469.60,16446.0,United Kingdom
540422,C581484,23843,"PAPER CRAFT , LITTLE BIRDIE",-80995,2.08,-168469.60,16446.0,United Kingdom
61619,541431,23166,MEDIUM CERAMIC TOP STORAGE JAR,74215,1.04,77183.60,12346.0,United Kingdom
61624,C541433,23166,MEDIUM CERAMIC TOP STORAGE JAR,-74215,1.04,-77183.60,12346.0,United Kingdom
222671,556444,22502,PICNIC BASKET WICKER 60 PIECES,60,649.50,38970.00,15098.0,United Kingdom
222681,C556445,M,Manual,-1,38970.00,-38970.00,15098.0,United Kingdom
524602,C580605,AMAZONFEE,AMAZON FEE,-1,17836.46,-17836.46,NaN,United Kingdom
43702,C540117,AMAZONFEE,AMAZON FEE,-1,16888.02,-16888.02,NaN,United Kingdom
43703,C540118,AMAZONFEE,AMAZON FEE,-1,16453.71,-16453.71,NaN,United Kingdom
15016,C537630,AMAZONFEE,AMAZON FEE,-1,13541.33,-13541.33,NaN,United Kingdom


## 2.2.14 Revenue and Transaction Value Investigation

A preliminary transaction-line value was calculated as:

`LineValue = Quantity × Price`

This measure is used only for Data Understanding at this stage. It is not
yet treated as business revenue because the raw dataset contains multiple
transaction types, including merchandise sales, returns, cancellations,
fees, adjustments and other operational records.

### Raw Line-Value Summary

For the `Year 2009-2010` worksheet:

* Total raw line value: £9,539,484.63
* Positive line-value records: 511,566
* Negative line-value records: 10,208
* Zero line-value records: 3,687
* Minimum line value: -£53,594.36
* Maximum line value: £25,111.09

For the `Year 2010-2011` worksheet:

* Total raw line value: £9,747,765.93
* Positive line-value records: 530,105
* Negative line-value records: 9,290
* Zero line-value records: 2,515
* Minimum line value: -£168,469.60
* Maximum line value: £168,469.60

### Extreme-Value Findings

The largest absolute values are associated with several different types
of records.

Examples include:

* `Adjust bad debt` records with large negative prices.
* `Manual` transactions with unusually large prices.
* `AMAZON FEE` records with large transaction values.
* `BANK CHARGES` records.
* Merchandise transactions containing unusually large quantities.

The 2010-2011 worksheet contains particularly large quantity values,
including a transaction for 80,995 units of `PAPER CRAFT , LITTLE BIRDIE`
and a corresponding cancellation record for -80,995 units.

Another example contains 74,215 units of `MEDIUM CERAMIC TOP STORAGE JAR`
and a corresponding cancellation record for -74,215 units.

These observations demonstrate that unusually large values cannot be
classified as invalid solely because they are statistical outliers.

### Key Finding

`Quantity × Price` provides a useful raw transaction-value measure, but
it cannot yet be interpreted directly as sales revenue.

Negative and zero values have multiple possible business meanings, and
large positive values may represent legitimate merchandise transactions,
manual records, fees or other operational activity.

### Data-Quality Implications

The final reporting pipeline must distinguish transaction types before
calculating business KPIs such as sales revenue, returns and net revenue.

Transaction classification should consider multiple attributes,
including:

* `Invoice`
* `StockCode`
* `Description`
* `Quantity`
* `Price`
* `Customer ID`

No records are removed solely because their line value is negative, zero,
or unusually large during the Data Understanding phase.


In [8]:
for sheet in excel_file.sheet_names:
    data = pd.read_excel(
        raw_file,
        sheet_name=sheet
    )

    print(f"\n{sheet}")
    print("=" * 70)

    # Find exact duplicate rows
    duplicates = data[data.duplicated(keep=False)].copy()

    print("Total exact duplicate rows:", len(duplicates))
    print("Unique duplicated row patterns:", duplicates.drop_duplicates().shape[0])

    print("\nTop duplicated row patterns:")
    display(
        duplicates
        .value_counts()
        .head(15)
        .reset_index(name="Duplicate_Count")
    )

    # How many duplicate groups have more than 2 copies?
    duplicate_groups = (
        duplicates
        .groupby(list(data.columns))
        .size()
        .reset_index(name="Duplicate_Count")
    )

    print(
        "\nDuplicate groups occurring more than twice:",
        (duplicate_groups["Duplicate_Count"] > 2).sum()
    )

    print("\nLargest duplicate groups:")
    display(
        duplicate_groups
        .sort_values("Duplicate_Count", ascending=False)
        .head(10)
    )


Year 2009-2010
Total exact duplicate rows: 13283
Unique duplicated row patterns: 6418

Top duplicated row patterns:


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Duplicate_Count
0,534219,35953,FOLKART STAR CHRISTMAS DECORATIONS,1,2010-11-21 16:13:00,1.25,13230.0,United Kingdom,6
1,502660,17021,NAMASTE SWAGAT INCENSE,6,2010-03-25 17:18:00,0.30,13187.0,United Kingdom,6
2,496431,84826,ASSTD DESIGN 3D PAPER STICKERS,1,2010-02-01 12:30:00,0.85,16415.0,United Kingdom,6
3,525065,20894,HANGING BAUBLE T-LIGHT HOLDER LARGE,1,2010-10-03 14:28:00,2.95,16799.0,United Kingdom,6
4,501425,85040B,SET/4 BLUE FLOWER CANDLES IN BOWL,1,2010-03-16 12:00:00,1.65,17774.0,United Kingdom,5
5,492784,22114,HOT WATER BOTTLE TEA AND SYMPATHY,1,2009-12-18 17:24:00,3.95,17841.0,United Kingdom,5
6,537224,70007,HI TEC ALPINE HAND WARMER,1,2010-12-05 16:24:00,1.65,13174.0,United Kingdom,5
7,532679,22264,FELT FARM ANIMAL WHITE BUNNY,1,2010-11-14 12:17:00,0.85,16960.0,United Kingdom,5
8,530953,21231,SWEETHEART CERAMIC TRINKET BOX,4,2010-11-04 18:42:00,1.25,18168.0,United Kingdom,5
9,520570,21754,HOME BUILDING BLOCK WORD,1,2010-08-26 20:07:00,5.95,16549.0,United Kingdom,5



Duplicate groups occurring more than twice: 375

Largest duplicate groups:


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Duplicate_Count
5629,534219,35953,FOLKART STAR CHRISTMAS DECORATIONS,1,2010-11-21 16:13:00,1.25,13230.0,United Kingdom,6
1431,502660,17021,NAMASTE SWAGAT INCENSE,6,2010-03-25 17:18:00,0.30,13187.0,United Kingdom,6
757,496431,84826,ASSTD DESIGN 3D PAPER STICKERS,1,2010-02-01 12:30:00,0.85,16415.0,United Kingdom,6
3995,525065,20894,HANGING BAUBLE T-LIGHT HOLDER LARGE,1,2010-10-03 14:28:00,2.95,16799.0,United Kingdom,6
1242,501425,85040B,SET/4 BLUE FLOWER CANDLES IN BOWL,1,2010-03-16 12:00:00,1.65,17774.0,United Kingdom,5
429,492784,22114,HOT WATER BOTTLE TEA AND SYMPATHY,1,2009-12-18 17:24:00,3.95,17841.0,United Kingdom,5
6200,537224,70007,HI TEC ALPINE HAND WARMER,1,2010-12-05 16:24:00,1.65,13174.0,United Kingdom,5
5204,532679,22264,FELT FARM ANIMAL WHITE BUNNY,1,2010-11-14 12:17:00,0.85,16960.0,United Kingdom,5
4869,530953,21231,SWEETHEART CERAMIC TRINKET BOX,4,2010-11-04 18:42:00,1.25,18168.0,United Kingdom,5
3473,520570,21754,HOME BUILDING BLOCK WORD,1,2010-08-26 20:07:00,5.95,16549.0,United Kingdom,5



Year 2010-2011
Total exact duplicate rows: 10147
Unique duplicated row patterns: 4879

Top duplicated row patterns:


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Duplicate_Count
0,555524,22698,PINK REGENCY TEACUP AND SAUCER,1,2011-06-05 11:37:00,2.95,16923.0,United Kingdom,20
1,555524,22697,GREEN REGENCY TEACUP AND SAUCER,1,2011-06-05 11:37:00,2.95,16923.0,United Kingdom,12
2,572861,22775,PURPLE DRAWERKNOB ACRYLIC EDWARDIAN,12,2011-10-26 12:46:00,1.25,14102.0,United Kingdom,8
3,572344,M,Manual,48,2011-10-24 10:43:00,1.50,14607.0,United Kingdom,6
4,540524,21756,BATH BUILDING BLOCK WORD,1,2011-01-09 12:53:00,5.95,16735.0,United Kingdom,6
5,541266,21754,HOME BUILDING BLOCK WORD,1,2011-01-16 16:25:00,5.95,15673.0,United Kingdom,6
6,541266,21755,LOVE BUILDING BLOCK WORD,1,2011-01-16 16:25:00,5.95,15673.0,United Kingdom,6
7,538514,21756,BATH BUILDING BLOCK WORD,1,2010-12-12 14:27:00,5.95,15044.0,United Kingdom,6
8,578289,23395,BELLE JARDINIERE CUSHION COVER,1,2011-11-23 14:07:00,3.75,17841.0,United Kingdom,6
9,550331,85202,HANGING WOOD AND FELT HEART,1,2011-04-17 13:43:00,0.42,15555.0,United Kingdom,5



Duplicate groups occurring more than twice: 290

Largest duplicate groups:


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Duplicate_Count
1596,555524,22698,PINK REGENCY TEACUP AND SAUCER,1,2011-06-05 11:37:00,2.95,16923.0,United Kingdom,20
1595,555524,22697,GREEN REGENCY TEACUP AND SAUCER,1,2011-06-05 11:37:00,2.95,16923.0,United Kingdom,12
3171,572861,22775,PURPLE DRAWERKNOB ACRYLIC EDWARDIAN,12,2011-10-26 12:46:00,1.25,14102.0,United Kingdom,8
3115,572344,M,Manual,48,2011-10-24 10:43:00,1.50,14607.0,United Kingdom,6
470,540524,21756,BATH BUILDING BLOCK WORD,1,2011-01-09 12:53:00,5.95,16735.0,United Kingdom,6
524,541266,21754,HOME BUILDING BLOCK WORD,1,2011-01-16 16:25:00,5.95,15673.0,United Kingdom,6
525,541266,21755,LOVE BUILDING BLOCK WORD,1,2011-01-16 16:25:00,5.95,15673.0,United Kingdom,6
346,538514,21756,BATH BUILDING BLOCK WORD,1,2010-12-12 14:27:00,5.95,15044.0,United Kingdom,6
4212,578289,23395,BELLE JARDINIERE CUSHION COVER,1,2011-11-23 14:07:00,3.75,17841.0,United Kingdom,6
1289,550331,85202,HANGING WOOD AND FELT HEART,1,2011-04-17 13:43:00,0.42,15555.0,United Kingdom,5
